<a href="https://colab.research.google.com/github/OlegV12/Stats/blob/Lesson_2/mathStats.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from math import factorial, pow, e

In [37]:

def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [3]:
def bernulli(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  return combinations(n, k) * pow(p, k) * pow((1-p), (n-k))

In [50]:
def poisson(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  a = n * p # lambda

  return pow(a, k) / factorial(k) * pow(e, (-a))



Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

In [36]:
result = bernulli(100, 85, 0.8)
result

0.048061793700746355

Вероятность того, что лампочка перегорит в течение первого дня эксплуатации, равна 0.0004. В жилом комплексе после ремонта в один день включили 5000 новых лампочек. Какова вероятность, что ни одна из них не перегорит в первый день? Какова вероятность, что перегорят ровно две?

In [51]:
result = poisson(5000, 0, 0.0004)
result

0.1353352832366127

Ответ: ни одна не перегорит с вероятностью 0.13528114551440706

In [52]:
result = poisson(5000, 2, 0.0004)
result

0.2706705664732254

Ответ: перегорят две с вероятностью 0.2706705664732254

Монету подбросили 144 раза. Какова вероятность, что орел выпадет ровно 70 раз?

In [53]:
result = bernulli(144, 70, 0.5)
result

0.06281178035144776

Ответ: вероятность выпадения орла 0.06281178035144776

В первом ящике находится 10 мячей, из которых 7 - белые. Во втором ящике - 11 мячей, из которых 9 белых. Из каждого ящика вытаскивают случайным образом по два мяча. Какова вероятность того, что все мячи белые? Какова вероятность того, что ровно два мяча белые? Какова вероятность того, что хотя бы один мяч белый?

In [15]:
# all white balls from 1st basket
all_combinations_1 = combinations(10, 2)
white_balls_combinations_1 = combinations(7, 2)
result_1 = white_balls_combinations_1/all_combinations_1
result_1

0.4666666666666667

In [16]:
# all white balls from 2nd basket
all_combinations_2 = combinations(11, 2)
white_balls_combinations_2 = combinations(9, 2)
result_2 = white_balls_combinations_2/all_combinations_2
result_2

0.6545454545454545

In [18]:
result_1 * result_2

0.3054545454545455

Ответ: все мячи белые с вероятностью 0.3054545454545455

In [32]:
# 2 white balls from 1st basket and 0 white balls from 2nd basket
two_and_zero = result_1 * (combinations(2, 2) / all_combinations_2)

In [29]:
# 0 white balls from 1st basket and 2 white balls from 2nd basket
zero_and_two = (combinations(3, 2) / all_combinations_1) * result_2

In [31]:
# 1 white ball from 1st basket and 1 white ball from 2nd basket
one_and_one = (combinations(7, 1) / all_combinations_1) * (combinations(9, 1) / all_combinations_2)

0.025454545454545455

In [33]:
res = two_and_zero + zero_and_two + one_and_one
res

0.07757575757575758

Ответ: ровно 2 мяча белые с вероятностью 0.07757575757575758

In [36]:
# one from first basket
first = (combinations(7, 1) / all_combinations_1)

# one from second basket
second = (combinations(3, 2) / all_combinations_1)

at_least_one = first + second - (first * second)
at_least_one

0.21185185185185185

Ответ: хотя бы один мяч белый с вероятностью 0.21185185185185185